torchtext/datasets/text_classification.py

```diff
@@ -124,18 +136,47 @@ def _setup_datasets(dataset_name, root='.data', ngrams=2, vocab=None, include_un
             test_csv_path = fname
 
     if vocab is None:
-        logging.info('Building Vocab based on {}'.format(train_csv_path))
-        vocab = build_vocab_from_iterator(_csv_iterator(train_csv_path, ngrams))
+        logging.info('Building Vocab based on train data')
+        if dataset_name == 'IMDB':
+            data_iterator = _imdb_iterator('train', extracted_files, tokenizer, ngrams)
+        else:
+            data_iterator = _csv_iterator(train_csv_path, tokenizer, ngrams)
+        vocab = build_vocab_from_iterator(data_iterator)
     else:
         if not isinstance(vocab, Vocab):
             raise TypeError("Passed vocabulary is not of type Vocab")
     logging.info('Vocab has {} entries'.format(len(vocab)))
+
     logging.info('Creating training data')
-    train_data, train_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(train_csv_path, ngrams, yield_cls=True), include_unk)
+    if dataset_name == 'IMDB':
+        data_iterator = _imdb_iterator('train', extracted_files,
+                                       tokenizer, ngrams, yield_cls=True)
+    else:
+        data_iterator = _csv_iterator(train_csv_path, tokenizer, ngrams, yield_cls=True)
+    data_iter = _create_data_from_iterator(vocab, data_iterator, removed_tokens)
+    train_data = []
+    train_labels = []
+    for cls, tokens in data_iter:
+        train_data.append((torch.tensor(cls),
+                           torch.tensor([token_id for token_id in tokens])))
+        train_labels.append(cls)
+    train_labels = set(train_labels)
```

Comment:
Can you give an example of what is `train_labels` and `test_labels`? I was expecting them to be ordered.

Reply:
For classification problem, the `class` is saved in `train_data` as `cls` and used as target. Each data point has a `cls` and labels is actually a set of those `cls`